src/dataflow/src/decode/mod.rs

```diff
@@ -366,72 +363,56 @@ where
 
     let mut value_decoder = get_decoder(value_encoding, debug_name, operators, true, metrics);
 
-    // The Debezium deduplication and upsert logic rely on elements for the same key going to the same worker.
-    // Other decoders don't care; so we distribute things round-robin (i.e., by "position"), and
-    // fall back to arbitrarily hashing by value if the upstream didn't give us a position.
-    // TODO(petrosagg): this function should accept a pact from the caller
-    let use_key_contract = matches!(envelope, SourceEnvelope::Debezium(..));
-    let results = stream.unary_frontier(
-        Exchange::new(move |x: &SourceOutput<Option<Vec<u8>>, Option<Vec<u8>>>| {
-            if use_key_contract {
-                x.key.hashed()
-            } else if let Some(position) = x.position {
-                position.hashed()
-            } else {
-                x.value.hashed()
-            }
-        }),
-        &op_name,
-        move |_, _| {
-            move |input, output| {
-                let mut n_errors = 0;
-                let mut n_successes = 0;
-                input.for_each(|cap, data| {
-                    let mut session = output.session(&cap);
-                    for SourceOutput {
-                        key,
-                        value,
-                        position,
-                        upstream_time_millis,
-                        partition,
-                    } in data.iter()
-                    {
-                        let key = key_decoder
-                            .as_mut()
-                            .and_then(|decoder| try_decode(decoder, key.as_ref()));
-
-                        let value = try_decode(&mut value_decoder, value.as_ref());
-
-                        if matches!(&key, Some(Err(_))) || matches!(&value, Some(Err(_))) {
-                            n_errors += 1;
-                        } else if matches!(&value, Some(Ok(_))) {
-                            n_successes += 1;
-                        }
+    let results = stream.unary_frontier(Pipeline, &op_name, move |_, _| {
```

Comment:
Before, we where potentially better utilizing all workers for decoding work. We could still exchange by `position.hashed()` or `x.value.hashed()` to keep that property.

Reply:
I thought we were exchanging unconditionally before this step but I was misremembering and we reshuffle after the source pipeline: https://github.com/MaterializeInc/materialize/blob/v0.18.0/src/dataflow/src/render/sources.rs#L631-L633

I'll put it back